kernel/src/fs/utils/channel.rs

```diff
@@ -1,436 +1,219 @@
-        self.peer_end().pollee.notify(IoEvents::IN);
-        self.this_end().pollee.invalidate();
-
-        if written_len > 0 {
-            Ok(written_len)
-        } else {
-            return_errno_with_message!(Errno::EAGAIN, "the channel is full");
+    /// Creates with the [`Pollee`] and the shutdown status.
+    pub fn with_pollee(pollee: Pollee, is_shutdown: bool) -> Self {
+        Self {
+            pollee,
+            is_shutdown: AtomicBool::new(is_shutdown),
         }
     }
-}
 
-impl<T: Pod> Producer<T> {
-    /// Tries to push `item` to the channel.
+    /// Clones and returns the [`Pollee`].
     ///
-    /// - Returns `Ok(())` if successful.
-    /// - Returns `Err(EPIPE)` if the channel is shut down.
-    /// - Returns `Err(EAGAIN)` if the channel is full.
-    pub fn try_push(&self, item: T) -> core::result::Result<(), (Error, T)> {
-        if self.is_shutdown() {
-            let err = Error::with_message(Errno::EPIPE, "the channel is shut down");
-            return Err((err, item));
-        }
-
-        self.0.push(item).map_err(|item| {
-            let err = Error::with_message(Errno::EAGAIN, "the channel is full");
-            (err, item)
-        })?;
-        self.peer_end().pollee.notify(IoEvents::IN);
-        self.this_end().pollee.invalidate();
-
-        Ok(())
+    /// Do not use this method to perform cheap operations on the [`Pollee`] (e.g.,
+    /// [`Pollee::notify`]). Use the methods below, such as `read_with`/`write_with`, instead. This
+    /// method is deliberately designed to force the [`Pollee`] to be cloned to avoid such misuse.
+    pub fn cloned_pollee(&self) -> Pollee {
+        self.pollee.clone()
     }
 }
 
-impl<T> Drop for Producer<T> {
-    fn drop(&mut self) {
-        self.shutdown();
+impl Default for ChannelPollee {
+    fn default() -> Self {
+        Self::new()
     }
 }
 
-impl<T> Consumer<T> {
-    fn this_end(&self) -> &FifoInner<RbConsumer<T>> {
-        &self.0.common.consumer
-    }
-
-    fn peer_end(&self) -> &FifoInner<RbProducer<T>> {
-        &self.0.common.producer
+impl AsRef<ChannelPollee> for ChannelPollee {
+    fn as_ref(&self) -> &ChannelPollee {
+        self
     }
-
-    fn check_io_events(&self) -> IoEvents {
-        let this_end = self.this_end();
-        let rb = this_end.rb();
-
-        let mut events = IoEvents::empty();
-        if self.is_shutdown() {
-            events |= IoEvents::HUP;
-        }
-        if !rb.is_empty() {
-            events |= IoEvents::IN;
-        }
-        events
-    }
-
-    impl_common_methods_for_channel!();
 }
 
-impl Consumer<u8> {
-    /// Tries to read `buf` from the channel.
-    ///
-    /// - Returns `Ok(_)` with the number of bytes read if successful.
-    /// - Returns `Ok(0)` if the channel is shut down and there is no data left.
-    /// - Returns `Err(EAGAIN)` if the channel is empty.
+impl<T: AsRef<ChannelPollee>> Peered<T> {
+    /// Reads from the channel and updates the local/remote [`Pollee`]s.
```

Comment:
It's better not to refer to it as `channel` anymore since the `Channel` struct has been removed.
If you want to maintain the concept of `Channel`, consider adding module-level documentation to explain what a `Channel` is and describe its relationship with the structs in this module.

Alternatively, you could remove all references to `Channel` in the comments, and perhaps the entire file should be renamed (perhaps to `endpoint.rs`) as well.